### PR TITLE
JIT: Support loading Swift self register from SwiftSelf struct

### PR DESCRIPTION
**Pull request by amanasifkhalid**
Adds JIT support for loading the Swift self/context register value from the `SwiftSelf` struct argument passed to a Swift call. Also enables corresponding tests.

### CODE REVIEW

**Comment by amanasifkhalid** [comment]
cc @jakobbotsch @kotlarmilos @jkoritzinsky. Here's a snippet of what the codegen looks like on x64, where the self register is r13:
```
       mov      r13, qword ptr [rbp-0x48]
       mov      rax, 0x7FF8FF2F59B8      ; function address
       mov      qword ptr [rbp-0x90], rax
       lea      rax, G_M44653_IG07
       mov      qword ptr [rbp-0x78], rax
       mov      rax, qword ptr [rbp-0x58]
       lea      rcx, bword ptr [rbp-0xA0]
       mov      qword ptr [rax+0x10], rcx
       mov      rax, qword ptr [rbp-0x58]
       mov      byte  ptr [rax+0x0C], 0
       call     [SelfContextTests:getMagicNumber(System.Runtime.InteropServices.Swift.SwiftSelf):long]
```

And on arm64, where the self register is x20:
```
            ldr     x20, [fp, #0x80]	// [V01 loc1]
            movz    x0, #0x4988
            movk    x0, #0x2103 LSL #16
            movk    x0, #0x7FF9 LSL #32
            str     x0, [fp, #0x30]	// [V05 PInvokeFrame+0x18]
            adr     x0, [G_M44653_IG09]
            str     x0, [fp, #0x48]	// [V05 PInvokeFrame+0x30]
            ldr     x0, [fp, #0x70]	// [V04 FramesRoot]
            add     x1, fp, #32	// [V05 PInvokeFrame+0x08]
            str     x1, [x0, #0x10]
            ldr     x0, [fp, #0x70]	// [V04 FramesRoot]
            strb    wzr, [x0, #0x0C]
            movz    x0, #0x49C8      // code for SelfContextTests:getMagicNumber(System.Runtime.InteropServices.Swift.SwiftSelf):long
            movk    x0, #0x2103 LSL #16
            movk    x0, #0x7FF9 LSL #32
            ldr     x0, [x0]
            blr     x0
```

File: src/coreclr/jit/morph.cpp [hunk 1]
```diff
@@ -3179,7 +3181,9 @@ GenTreeCall* Compiler::fgMorphArgs(GenTreeCall* call)
         }
 
         // TODO-ARGS: Review this, is it really necessary to treat them specially here?
-        if (call->gtArgs.IsNonStandard(this, call, &arg) && arg.AbiInfo.IsPassedInRegisters())
+        // Exception: Lower SwiftSelf struct arg to GT_LCL_FLD
+        if (call->gtArgs.IsNonStandard(this, call, &arg) && arg.AbiInfo.IsPassedInRegisters() &&
+            (arg.GetWellKnownArg() != WellKnownArg::SwiftSelf))
         {
```
Comment (jakobbotsch) [inline src/coreclr/jit/morph.cpp:3187]: Why is this necessary? That doesn't seem right.

Comment (jakobbotsch) [reply src/coreclr/jit/morph.cpp:3187]: Likely we should instead insert a new arg in the importer that represents "self" as a primitive instead of keeping this as a struct argument. It being a struct argument is just an artifact of the design and it wouldn't work to keep it as a struct on all ABIs.

Comment (jakobbotsch) [reply src/coreclr/jit/morph.cpp:3187]: Let's just keep it as is for now. I'm refactoring stuff for the struct work which should make this change simple to make.

**Review by jakobbotsch — APPROVED** [review]

File: src/coreclr/jit/gentree.cpp [hunk 2]
```diff
@@ -13091,6 +13098,8 @@ const char* Compiler::gtGetWellKnownArgNameForArgMsg(WellKnownArg arg)
         case WellKnownArg::ValidateIndirectCallTarget:
         case WellKnownArg::DispatchIndirectCallTarget:
             return "cfg tgt";
+        case WellKnownArg::SwiftSelf:
+            return "swift self";
```
Comment (jakobbotsch) [inline src/coreclr/jit/gentree.cpp:13102]: This should also be added to `getWellKnownArgName`.

Comment (jakobbotsch) [reply src/coreclr/jit/gentree.cpp:13102]: Ah sorry, I missed that you already added it there as well :)

File: src/coreclr/jit/gentree.h [hunk 3]
```diff
@@ -4401,7 +4401,7 @@ enum class CFGCallKind
 
 class CallArgs;
 
-enum class WellKnownArg
+enum class WellKnownArg : unsigned
```
Comment (jakobbotsch) [inline src/coreclr/jit/gentree.h:4404]: Why this change? (Seems fine, just curious)

File: src/coreclr/jit/morph.cpp (hunk 1 above)
Comment (amanasifkhalid) [reply src/coreclr/jit/morph.cpp:3187]: > Likely we should instead insert a new arg in the importer that represents "self" as a primitive instead of keeping this as a struct argument.

Yeah, the fact that the argument was being kept as `TYP_STRUCT` was causing asserts in later phases.

> Let's just keep it as is for now. I'm refactoring stuff for the struct work which should make this change simple to make.

Sure thing.

File: src/coreclr/jit/gentree.h (hunk 3 above)
Comment (amanasifkhalid) [reply src/coreclr/jit/gentree.h:4404]: `CallArg::m_wellKnownArg` is 5 bits, and `WellKnownArg::SwiftSelf` has a value of 16 (also 5 bits). If `m_wellKnownArg` is signed, then we lose the most-significant bit, so comparisons like `m_wellKnownArg == WellKnownArg::SwiftSelf` fail.

**Comment by amanasifkhalid** [comment]
ryujit-bot doesn't seem to be running, so I'll point out here that this is a [no-diff](https://dev.azure.com/dnceng-public/public/_build/results?buildId=584950&view=ms.vss-build-web.run-extensions-tab) change.
